api/src/middleware/authorization-middleware.ts

```diff
@@ -0,0 +1,102 @@
+import { type NextFunction, type Response } from "express"
+import { type Request as JwtRequest } from "express-jwt"
+import { isNil } from "lodash"
+
+import logger from "@/utils/logger"
+import { User } from "@/models"
+
+import auth0Integration, { Auth0PayloadError } from "@/integrations/auth0-integration"
+
+export type AuthorizationRequest = JwtRequest & {
+  user?: User
+}
+
+const USER_CACHE_EXPIRATION_MS = 60 * 1000 // 1 minutes
+
+interface CachedUser {
+  promise: Promise<User>
+  timestamp: number
+}
+
+const userDebounceCache = new Map<string, CachedUser>()
+
+function cleanUpCache() {
+  const now = Date.now()
+  for (const [key, cached] of userDebounceCache.entries()) {
+    if (now - cached.timestamp > USER_CACHE_EXPIRATION_MS) {
+      userDebounceCache.delete(key)
+    }
+  }
+}
+
+export async function ensureUserFromAuth0Token(token: string): Promise<User> {
+  const { auth0Subject, email, firstName, lastName } = await auth0Integration.getUserInfo(token)
+  const user = await User.findOne({ where: { sub: auth0Subject } })
+
+  if (!isNil(user)) {
+    return user
+  }
+
+  const existingUser = await User.findOne({
+    where: { sub: auth0Subject },
+  })
+
+  if (existingUser) {
+    return existingUser
+  }
+
+  const newUser = await User.create({
+    sub: auth0Subject,
+    email: email,
+    firstName: firstName,
+    lastName: lastName,
+    roles: [User.Roles.USER],
+    status: User.Statuses.ACTIVE,
+  })
+
+  logger.info(`CREATED USER FOR ${email}: ${JSON.stringify(newUser.dataValues)}`)
+  return newUser
+}
+
+export async function authorizationMiddleware(
+  req: AuthorizationRequest,
+  res: Response,
+  next: NextFunction
+) {
+  const token = req.headers.authorization || ""
+
+  cleanUpCache()
```

Comment:
Having the cache cleanup here unnecessarily slows down unrelated requests, vs. using a cache with a builtin timeout like node-cache.
Since this middleware is running before _every_ request, we want it to be as fast as possible.